### ISSUE TITLE
Document how to set environment variables in subprocesses

### ISSUE THREAD

**Issue by martine** [martine/ninja#1002]
See discussion on
https://groups.google.com/d/topic/ninja-build/ZGZ2Ewxsxaw/discussion
and previous thread linked there.

**Comment by nirbheek** [comment]
Reading through that discussion, it seems that environment variables aren't supported in Ninja because `CreateProcess` doesn't support setting `PATH` find the executable to run when using `lpCommandLine` instead of `lpApplicationName`? The thing is, `execvpe` has the same restriction, so I don't understand what the issue is here.

It's really simple to set the environment **inherited** by a process on all platforms, so I'd like to understand what the blocker is here. Cheers!

**Comment by nirbheek** [comment]
FWIW, we would really make good use of this feature in Meson: https://github.com/mesonbuild/meson/issues/266, https://github.com/mesonbuild/meson/issues/384, etc.

**Comment by nirbheek** [comment]
Would the project be interested in a PR that implements this? It's a troublesome blocker downstream for https://github.com/mesonbuild/meson/ since we have to add wrapper scripts for any command that needs environment variables set, which is slowly turning out to be quite a few.

Even Makefiles and Visual Studio project files support this, so Ninja stands out in not supporting this. :)

We'd be happy to implement this though if there is interest in having this feature in Ninja.

**Comment by nico** [comment]
On POSIXy platforms, you can do
```
rule foo
  command = ENV1=env1 ENV2=env2 my_command
```
or
```
rule foo
  command = $env my_command
foo out: in
  env = ENV1=env1 ENV2=env2
```

Is that not enough for you? Most commands shouldn't need env vars, and generally supporting less stuff keeps ninja simple.

**Comment by nirbheek** [comment]
Yes, that part [is documented](https://ninja-build.org/manual.html#ref_rule_command), but Meson is a cross-platform build system (meant to have feature-parity across all supported platforms) and we currently have to use Python script wrappers whenever we need to set environment variables or have arguments with special characters (especially newlines).

At first, we also thought that setting env vars is a niche use-case and pushed back when users asked for it, but a surprisingly large number of people need it. For instance, gobject-introspection in GNOME requires you to set `CC`/`CXX`/etc while invoking its tools, otherwise it uses an auto-detection mechanism that ends up using the wrong compiler. "Custom targets" in Meson can run arbitrary commands, and people often show up asking how they can set environment variables while running some tool that they can't change the behaviour of.

The Python script workaround is ok, but invoking the python interpreter for each command really slows things down sometimes.

**Comment by nico** [comment]
Maybe gobject-introspection could accept those via command line args instead? (In my experience, builds that rely on env vars instead of flags tend to be much more vulnerable to devs screwing up their local env, etc.)

MSVC's cl.exe requires a few env vars to find includes and whatnot; for that ninja has `ninja -t msvc` which has a `-e` flag that can load an environment from a file. So meson could theoretically use that on Windows, and the posix stuff elsewhere.

But using python wrappers for the rare, "needs to be flexible" case and making the common commands work without reliance on the env seems like the best approach to me.

**Comment by sam-github** [comment]
I think I've run into this issue. I'm trying to build a FIPS executable on window, and need to follow these directions:

https://www.openssl.org/docs/fips/UserGuide-2.0.pdf, section 5.3.2:

```
For the Windows®
 environment a perl script fipslink.pl is provided which performs a
function similar to fipsld for Unix®
/Linux®
. Several environment variables need to be set:
FIPS_LINK is the linker name, normally “link”
FIPS_CC is the C compiler name, normally “cl”
FIPS_CC_ARGS is a string of C compiler arguments for compiling fips_premain.c
PREMAIN_DSO_EXE should be set to the path to fips_premain_dso.exe if a DLL is
being linked (can be omitted otherwise)
PREMAIN_SHA1_EXE is the full path to fips_standalone_sha1.exe
FIPS_TARGET is the path of the target executable or DLL file
```

Even though FIPS_CC is defined in my environment, it looks like ninja is not passing the env to the child process (`perl fipslink.pl ...`). I'll see if -e works for me.

**Comment by nirbheek** [comment]
Another use-case for this is setting `LD_LIBRARY_PATH` before running executables built in the project. This is often necessary with executables linked to shared libraries built in the project to ensure that already-existing versions in a custom prefix are not used (`-Wl,-rpath` [doesn't work](https://stackoverflow.com/a/33520976) on distros that set `DT_RUNPATH` instead of `DT_RPATH` such as Debian and Ubuntu). Autotools (libtool) uses shell scripts for this. It would be a significant slowdown of the build if we had to spawn a Python interpreter or run a shell script for such cases.

As another datapoint, cmake came across this use-case so much that they have [a C wrapper for it](https://gitlab.kitware.com/cmake/cmake/blob/master/Source/kwsys/SharedForward.h.in) which sets the environment and then calls the process. They need it anyway for their make backend, but ninja-only build systems would greatly benefit from this feature.

**Comment by evmar** [comment]
I'm not really sure what this bug is discussing anymore, since it looks like it was originally about updating the docs. :)

For `LD_LIBRARY_PATH`, setting it via the command line already works as Nico mentioned above.  It doesn't involve another process; we must use a shell to parse the command line anyway.  (Subprocesses on Linux are oh-so-cheap anyway, you can't even measure `/bin/sh -c "echo hello"` because it is so fast.)

On Windows we have `ninja -t msvc`.  You might try writing your meson feature using that and then once we have some experience with that we could figure out how to improve it.

**Comment by marc-h38** [comment]
Kconfiglib has a hard `$srctree` requirement for out of source builds. It uses a few other variables:
https://github.com/ulfalizer/Kconfiglib/blob/424d0d38e7/kconfiglib.py#L108

> we currently have to use Python script wrappers whenever we need to set environment variables

From: https://github.com/ulfalizer/Kconfiglib/tree/424d0d38e7be15c5#overview
> The entire library is contained in kconfiglib.py. The bundled scripts are implemented on top of it. Implementing your own scripts should be relatively easy, if needed.

**Comment by nazar-pc** [comment]
I think I might have hit this in #1997 where Meson calls Ninja, Ninja calls Meson back, but without inherited environment variables and everything falls apart because of that.

**Comment by Hi-Angel** [comment]
So… Doesn't this feature belong to the goals of Ninja project? The [documentation says](https://ninja-build.org/manual.html#_design_goals), quoting:

> * very fast (i.e., instant) incremental builds, even for very large projects
> […]
> * when convenience and speed are in conflict, prefer speed.

From reading the discussion, supporting this feature will improve build speed in many cases. So, is there any reason not to support it?

**Comment by wzssyqa** [comment]
I am not sure is it for my use case: run unittest of llvm with qemu-user.

Background:
To run a dynlink app with qemu user, we should set env `QEMU_LD_PREFIX`, for example
```
QEMU_LD_PREFIX=/usr/mipsel-linux-gnu
```

While `ninja check` still complains that:
```
mipsel-binfmt-P: Could not open '/lib/ld.so.1': No such file or directory
```

It means that `QEMU_LD_PREFIX` is not passed to subprocess.

**Comment by vipcxj** [comment]
Almost 10 years later, are environment variables such a horrible thing?

**Comment by marc-h38** [comment]
> Almost 10 years later, are environment variables such a horrible thing?

No, they aged well - like wine!

More seriously: yes they are still horrible for the simple reason that Global Variables are horrible. Environment variables are even worse because they span multiple processes.

BUT, a build very frequently involves tools that you "can't change the behaviour of". Some toolchains need them, example: https://github.com/thesofproject/sof/commit/7683707944194155dfcb06700ff1c865773e272e

**Comment by eli-schwartz** [comment]
> More seriously: yes they are still horrible for the simple reason that Global Variables are horrible. Environment variables are even worse because they span multiple processes.

Once again, this is absolute nonsense.

Everyone uses CFLAGS and LDFLAGS, for example! They are parsed by cmake, meson, autotools, configure.py, and Github Actions. Environment variables as they are used in all scenarios for which "environment variable hardliners" hate them, are always carefully constrained and scoped and get set immediately and unconditionally for the sole purpose of overriding behavior within the same process flow.

In this case the explicit purpose of setting environment variables in a ninja build rule is so that you can statically set an environment variable to a static value that is scoped to a single subprocess. The objections against this are baffling.

**Comment by marc-h38** [comment]
> Everyone uses CFLAGS and LDFLAGS, for example!

Yes, and everyone also uses CCFLAGS, EXTRA_CFLAGS, KCFLAGS, KBUILD_LDFLAGS, EXTRA_LDLIBS, HOSTCFLAGS, USERCXXFLAGS, CFLAGS_<filename>.o, REALCFLAGS, CFLAGS_INTERNAL and what not. That's Global Variables for you.

From https://www.gnu.org/software/make/manual/html_node/Override-Directive.html
> The override directive was not invented for escalation in **the war between makefiles and command arguments**.

----------

> Environment variables as they are used in all scenarios for which "environment variable hardliners" hate them, are always carefully constrained ...

ROFL. In other, breaking news, C/C++ toolchains are and have always been a walk in the park - especially when cross-compiling.

To be clear: I do NOT know any silver bullet. I don't "hate" any necessary evil either.

**Comment by vipcxj** [comment]
In fact, as cloud services and containerization spread, environment variables are only going to get better. Because of the environmental isolation that comes with containerization, many of the shortcomings of environment variables have been greatly improved. Times are changing.
